### PR TITLE
[Improvement-15603][API] When removing or modifying a workflow the system can check if there any tasks depend on it.

### LINKED ISSUE #15603: [Improvement][API] When removing or modifying a workflow the system should check if there any tasks depend on it.
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

When removing or modifying a workflow the backend api doesn't check if there any tasks depend on it. If you remove a workflow,  its dependent tasks  can't execute properly. Likewise, if you remove a task. So the backend api needs to check.

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by calvinjiang**
## Purpose of the pull request

This PR will close #15603 .

## Verify this pull request
This change added tests and can be verified as follows:
- *Manually verified the change by testing locally.* 

The test results are as follows:

image
image

image
image

image

image
image

### CODE REVIEW

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/WorkFlowLineageController.java [hunk 1]
```diff
@@ -135,4 +135,26 @@ public Result verifyTaskCanDelete(@Parameter(hidden = true) @RequestAttribute(va
...
+                                                                        @RequestParam(value = "workFlowCode") Long workFlowCode,
+                                                                        @RequestParam(value = "taskCode", required = false, defaultValue = "0") Long taskCode) {
+        try {
+            Map<String, Object> result =
+                    workFlowLineageService.queryDownstreamDependentTasks(projectCode, workFlowCode, taskCode);
+            return returnDataList(result);
+        } catch (Exception e) {
+            log.error(QUERY_WORKFLOW_LINEAGE_ERROR.getMsg(), e);
+            return error(QUERY_WORKFLOW_LINEAGE_ERROR.getCode(), QUERY_WORKFLOW_LINEAGE_ERROR.getMsg());
+        }
```
Comment (zhuangchong) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/WorkFlowLineageController.java]: I recommend using `@ApiException` annotation for this part of the code.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java [hunk 2]
```diff
@@ -815,6 +815,13 @@ public Map<String, Object> updateProcessDefinition(User loginUser,
         return result;
     }
 
+    /**
+     * Check if there're any tasks that depend on the process
+     */
+    private void checkIfDependentTaskExists(ProcessDefinition processDefinition) {
```
Comment (zhuangchong) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java]: no called.

File: dolphinscheduler-ui/src/views/projects/workflow/components/dag/dag-context-menu.tsx [hunk 3]
```diff
@@ -76,7 +81,19 @@ export default defineComponent({
...
     const projectCode = Number(route.params.projectCode)
+    const workflowCode = Number(route.params.code)
+    const { t } = useI18n()
+
+    // const dependenciesData = reactive({
+    //       showRef: false,
+    //       taskLinks: ref([]),
+    //       required: ref(false),
+    //       tip: ref(''), action: () => {}
+    // })
```
Comment (zhuangchong) [inline dolphinscheduler-ui/src/views/projects/workflow/components/dag/dag-context-menu.tsx]: Please delete this part.

**Review by songjianet — CHANGES_REQUESTED** [review]
Please delete introduced but unused variables

**Review by zhuangchong — DISMISSED** [review]
+1

**Review by songjianet — CHANGES_REQUESTED** [review]
image

**Review by songjianet — APPROVED** [review]

**Comment by qingwli** [comment]
Looks like check https://github.com/apache/dolphinscheduler/actions/runs/8215624741/job/22469292085 not pass @calvinjiang  @songjianet

**Comment by SbloodyS** [comment]
> Looks like check https://github.com/apache/dolphinscheduler/actions/runs/8215624741/job/22469292085 not pass @calvinjiang @songjianet

It's odd that `dolphinscheduler-api` module build passed...

**Comment by SbloodyS** [comment]
Maybe the command `npm install pnpm@7 -g` install an incompatible version of `pnpm` in
https://github.com/apache/dolphinscheduler/blob/2395ba5ef84bdbce127edc36e6dd7e4f064ff0b1/.github/workflows/frontend.yml#L60-L64

PTAL @songjianet

**Comment by qingwli** [comment]
Looks like some code format issue

**Pull request merged**